File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> res;

    bool isPalindrome(string s) {
        string t = s;
        reverse(t.begin(), t.end());
        return s == t;
    }

    void myrec(int idx, int n, string s, vector<string>& palindrome) {
        if (idx == n) {
            res.push_back(palindrome);
            return;
        }
        string temp = "";
        for (int i = idx; i < n; i++) {
            temp += s[i];
            if (isPalindrome(temp)) {
                palindrome.push_back(temp);
                myrec(i + 1, n, s, palindrome);
                palindrome.pop_back();
            }
        }
    }

    vector<vector<string>> partition(string s) {
        vector<string> palindrome;
        myrec(0, s.length(), s, palindrome);
        return res;
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    public static boolean ispalindrome(String s){
        int n=s.length();
        int i=0;
        while(i<n){
            if(s.charAt(i)!=s.charAt(n-i-1)){
                return false;
            }
            i++;
        }
        return true;
    }
    public static void myrec(List<List<String>> res,List<String> temp,int n,int idx,String s){
        //base condition
        if(idx==n){
            res.add(new ArrayList<>(temp));
            return;
        }
        //recursive calls
        String s2="";
        for(int i=idx;i<n;i++){
            s2+=s.charAt(i);
            if(ispalindrome(s2)){
                temp.add(s2);
                myrec(res,temp,n,i+1,s);
                temp.remove(temp.size()-1);
            }
        }
    }
    public List<List<String>> partition(String s) {
        List<List<String>> res=new ArrayList<>();
        List<String> temp=new ArrayList<>();

        int n=s.length();
        myrec(res,temp,n,0,s);
        return res;
    }
}